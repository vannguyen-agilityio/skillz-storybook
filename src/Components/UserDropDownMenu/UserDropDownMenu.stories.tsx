import Storybook from '@storybook/react';

// Components
import UserDropDownMenu from '.';

// Code
import { asDefault } from './UserDropDownMenu.code';

const USER_INFO = {
  id: '1',
  fullName: 'Dehlela Shabir',
  job: 'Instructor',
  profileImage: 'https://gravatar.com/avatar/d7ce2a42de05bc0e7216db3570111d56?s=400&d=mp&r=x'
};

export default {
  title: 'Components/UserDropDownMenu',
  component: UserDropDownMenu,
  parameters: {
    componentSource: {
      code: asDefault,
      language: 'typescript'
    }
  }
} as Storybook.ComponentMeta<typeof UserDropDownMenu>;

const Template: Storybook.ComponentStory<typeof UserDropDownMenu> = (args) => {
  return <UserDropDownMenu {...args} />;
};

export const UserDropDownMenuDefault = Template.bind({});
UserDropDownMenuDefault.args = {
  userInfo: USER_INFO
};
