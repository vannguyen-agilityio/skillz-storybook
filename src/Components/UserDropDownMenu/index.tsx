import {
  Menu as MenuChakra,
  MenuButton,
  MenuGroup,
  IconButton,
  MenuList,
  MenuItem,
  MenuDivider,
  Text
} from '@chakra-ui/react';

// Components
import { Avatar } from '@components/index';

// Enums
import { SIZES } from '@enums';

// Interface
import { IUserDropDownMenu } from '@interfaces/userDropDown.interface';

export const UserDropDownMenu = ({ userInfo, ...rest }: IUserDropDownMenu) => {
  return (
    <MenuChakra
      placement='bottom-end'
      strategy='fixed'
      {...rest}
    >
      <MenuButton
        as={IconButton}
        border='none'
        aria-label='Options'
        variant='transparent'
        data-testid='userDropDownMenu'
        fontSize='sm'
        textAlign='left'
        leftIcon={
          <Avatar
            src={userInfo?.profileImage}
            size={SIZES.MD}
          />
        }
      >
        <Text
          color='black'
          fontSize='sm'
          lineHeight='base'
        >
          {userInfo?.fullName}
        </Text>
        <Text
          textTransform='uppercase'
          color='grey.700'
          lineHeight='base'
        >
          {userInfo?.job}
        </Text>
      </MenuButton>
      <MenuGroup>
        <MenuList
          color='grey.400'
          boxShadow='md'
        >
          <MenuItem
            as='a'
            href='#'
          >
            Profile &#38; account
          </MenuItem>
          <MenuDivider
            borderColor='grey.50'
            my='2.5'
          />
          <MenuItem onClick={() => {}}>Logout</MenuItem>
        </MenuList>
      </MenuGroup>
    </MenuChakra>
  );
};

export default UserDropDownMenu;
