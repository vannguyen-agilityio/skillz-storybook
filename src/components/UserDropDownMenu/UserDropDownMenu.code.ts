export const asDefault = `
import UserDropDownMenu from '.';

/**
 * An accessible dropdown menu for the common dropdown menu button design pattern. Menu uses roving tabIndex for focus management.
 *
 * @see Docs https://chakra-ui.com/docs/components/menu/
 */

 <UserDropDownMenu />
`;
