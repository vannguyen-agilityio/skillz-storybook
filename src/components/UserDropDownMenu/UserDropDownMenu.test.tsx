import { cleanup, render, screen } from '@testing-library/react';
import * as renderer from 'react-test-renderer';

import UserDropDownMenu from './index';

const USER_INFO = {
  id: '1',
  fullName: 'Dehlela Shabir',
  job: 'Instructor',
  profileImage: 'https://gravatar.com/avatar/d7ce2a42de05bc0e7216db3570111d56?s=400&d=mp&r=x'
};

describe('DropDownMenu render', () => {
  afterEach(cleanup);

  it('should render DropDownMenu component', () => {
    const { container } = render(<UserDropDownMenu>{}</UserDropDownMenu>);

    const userDropDownMenu = container.querySelector('.chakra-menu__menu-list');
    expect(userDropDownMenu).toBeTruthy();
  });

  it('should render check data-testid UserDropDownMenu component', () => {
    render(<UserDropDownMenu userInfo={USER_INFO}>{}</UserDropDownMenu>);

    const userDropDownMenu = screen.getByTestId('userDropDownMenu');
    expect(userDropDownMenu).toBeTruthy();
  });

  describe('jest Snapshot testing', () => {
    it('matches UserDropDownMenu Snapshot', () => {
      const userDropDownMenu = renderer.create(<UserDropDownMenu userInfo={USER_INFO}>{}</UserDropDownMenu>).toJSON();
      expect(userDropDownMenu).toMatchSnapshot();
    });
  });

  it('should render UserDropDownMenu component with userInfo props', () => {
    const { container } = render(<UserDropDownMenu userInfo={USER_INFO}>{}</UserDropDownMenu>);

    const userDropDownMenu = container.querySelector('.chakra-menu__menu-list');
    expect(userDropDownMenu).toBeTruthy();
  });
});
