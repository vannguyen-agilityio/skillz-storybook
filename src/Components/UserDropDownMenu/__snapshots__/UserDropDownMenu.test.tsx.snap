// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownMenu render jest Snapshot testing matches UserDropDownMenu Snapshot 1`] = `
[
  <button
    aria-controls="menu-list-:r1:"
    aria-expanded={false}
    aria-haspopup="menu"
    aria-label="Options"
    className="chakra-button chakra-menu__menu-button css-9mcn5g"
    data-testid="userDropDownMenu"
    id="menu-button-:r1:"
    onClick={[Function]}
    onKeyDown={[Function]}
    type="button"
  >
    <span
      className="chakra-button__icon css-1wh2kri"
    >
      <span
        aria-hidden={true}
        className="chakra-avatar css-1ruhx4w"
        focusable={false}
      >
        <svg
          aria-label=" avatar"
          className="chakra-avatar__svg css-16ite8i"
          role="img"
          viewBox="0 0 128 128"
        >
          <path
            d="M103,102.1388 C93.094,111.92 79.3504,118 64.1638,118 C48.8056,118 34.9294,111.768 25,101.7892 L25,95.2 C25,86.8096 31.981,80 40.6,80 L87.4,80 C96.019,80 103,86.8096 103,95.2 L103,102.1388 Z"
            fill="currentColor"
          />
          <path
            d="M63.9961647,24 C51.2938136,24 41,34.2938136 41,46.9961647 C41,59.7061864 51.2938136,70 63.9961647,70 C76.6985159,70 87,59.7061864 87,46.9961647 C87,34.2938136 76.6985159,24 63.9961647,24"
            fill="currentColor"
          />
        </svg>
      </span>
    </span>
    <span
      aria-hidden={true}
      className="css-xl71ch"
      focusable={false}
    >
      <p
        className="chakra-text css-pfrx0k"
      >
        Dehlela Shabir
      </p>
      <p
        className="chakra-text css-159d3cw"
      >
        Instructor
      </p>
    </span>
  </button>,
  <div
    className="chakra-menu__group"
    role="group"
  >
    <div
      className="css-0"
      style={
        {
          "inset": "0 auto auto 0",
          "minWidth": "max-content",
          "position": "fixed",
          "visibility": "hidden",
        }
      }
    >
      <div
        aria-orientation="vertical"
        className="chakra-menu__menu-list css-a45t2a"
        id="menu-list-:r1:"
        onKeyDown={[Function]}
        role="menu"
        style={
          {
            "opacity": 0,
            "transform": "scale(0.8) translateZ(0)",
            "transformOrigin": "var(--popper-transform-origin)",
            "visibility": "hidden",
          }
        }
        tabIndex={-1}
      >
        <a
          className="chakra-menu__menuitem css-1k756t9"
          href="#"
          id="menu-list-:r1:-menuitem-:r2:"
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          role="menuitem"
          tabIndex={0}
          type="button"
        >
          Profile & account
        </a>
        <hr
          aria-orientation="horizontal"
          className="chakra-menu__divider css-18a7dol"
        />
        <button
          className="chakra-menu__menuitem css-1k756t9"
          id="menu-list-:r1:-menuitem-:r3:"
          onClick={[Function]}
          onFocus={[Function]}
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
          onMouseMove={[Function]}
          role="menuitem"
          tabIndex={0}
          type="button"
        >
          Logout
        </button>
      </div>
    </div>
  </div>,
]
`;
